Completed precourse_2. Please review

sowmyadvn opened the pull request:
I have a question on quick sort swap method. I tried swapping with/without temp variable. When I did try to use in-place swap and the swapped values are equal, it resulted in 0's. 

Eg: For this array -> { 4, 3, 5, 2, 1, 3, 2, 3 }, the result I get for my in-place swap is {1, 2, 2, 0, 0, 3, 4, 5} . When I used swap with temp variable, result is {1, 2, 2, 3, 3, 3, 4, 5}. 

**Swap code:**
`void swap(int arr[],int i,int j){
        //swap values at indices i and j in-place  
        arr[i] = arr[i]+arr[j];
        arr[j] = arr[i]-arr[j];
        arr[i] = arr[i]-arr[j];
    }`

I tried figuring out but couldn't finally understand why that is happening. Am I missing something? Thanks!

super30admin commented:
Great Job Sowmya

In place swap won't work when we are trying to number by itself at same index. i.e when i and j are same and will give you a zero. A lot of times number is replaced by itself in the way we have written quick sort algo

1. Can you write the time complexities of all the solutions at top of the code?
2. Some follow ups:

- Can we do binary search without recursion? I mean using a while loop?

- where would you choose quicksort amd where merge sort?

- What is Big(O) of quick sort. Can you improve it?

sowmyadvn commented:
Thank you for following up. I just implemented iterative binary search and updated time complexities. After everyone is done submitting, can you give us more input on this question? 

> where would you choose quicksort amd where merge sort? 

Thanks!
